p2p/src/peers/mod.rs

```diff
@@ -179,6 +195,10 @@ impl Peers {
                     // If the index points to the same address that it is already
                     // in tried, we don't include in new bucket
                     if elem.is_none() || (elem.unwrap().address != address) {
+                        // If the source address that sent us this peer addresses is None, use the same address
+                        // that we want to add. This will make all the peer addresses that were added using manual methods
+                        // go to the same bucket that if it was announced by that address.
+                        let src_address = src_address.unwrap_or_else(|| address);
```

Comment:
This change may be dangerous, not sure. Perhaps we should replace `Option<SocketAddr>` with an new enum:

```rust
enum PeerSource {
    Config,
    JsonRpc,
    GetPeers(SocketAddr),
    InboundConnection(SocketAddr),
    IceBucket,
}
```

Reply:
At the end of the day, bucketing is only useful when the list of peers is not trusted. The peers in the ice bucket may or may not be trusted, but with the proposed change where `src_address = address`, 1 node sending us 100 peers will be the same as 100 nodes sending us 1 peer each.

Reply:
I'm not totally sure about this either. But I think that the most important property is that the position that a peer address takes in the buckets is deterministic. In other words, I believe that this change is of very little advantage for an attacker that may abuse the system so that many of their addresses go iced and then melt into `new`, because the bucket and index are still one-shot.

Reply:
@girazoki ?

Reply:
this actually breaks my PR I think haha. Still the most important thing is that from what I remember a source ip can only go to a subset of the buckets, so from that perspective with this change with a single attacker source address can fill more buckets. However, since attacker IPs will be propagated by potentially all peers in the network I do believe that this change might be of little advantage to him.